I would like to recommend that we change three elements of our coding standard

bobsummerwill opened the pull request:
I would like to recommend that we change three elements of our coding standard, which in my experience are rather poor recommendations.
1. Using tabs for formatting.   While this can be quite a religious issue, in my experience FORCING the use of tabs is a very poor practice, which causes lots of unnecessary work and problems.   Most software engineers work on multiple projects, with multiple editors and tools, across multiple operating systems.    The typical "well, just set up your tabs correctly in your editor" advice is inadequate and making the assumption that engineers are not switching between projects which likely have different standards.    Being reliant on some definition of "what a tab means" ends up impacting every diff tool and every editor for every team member.    Spaces just work everywhere.    I have had negative feedback from engineers outside our team on this exact point, and I agree with them 100%.    So does Greg.    Having this as a GOLDEN RULE for our coding standard casts a very poor light on the standard.    I think we should change it.    We don't need to "fix" all the code immediately.
2. Explicitly listing authors and maintainers PER FILE is unnecessary and egotistical.    It is a maintenance issue as well.    We should stop doing this, except where maintaining existing author information, or author information from files copied from other projects.
3. Leading underscores are, I believe, reserved for compiler and library authors.   Also, there should be no need for any prefixing if we are "m_" prefixing member variables - the most common source of naming clash.

bobsummerwill committed:
I would like to recommend that we change three elements of our coding standard, which in my experience are rather poor recommendations.

1. Using tabs for formatting.   While this can be quite a religious issue, in my experience FORCING the use of tabs is a very poor practice, which causes lots of unnecessary work and problems.   Most software engineers work on multiple projects, with multiple editors and tools, across multiple operating systems.    The typical "well, just set up your tabs correctly in your editor" advice is inadequate and making the assumption that engineers are not switching between projects which likely have different standards.    Being reliant on some definition of "what a tab means" ends up impacting every diff tool and every editor for every team member.    Spaces just work everywhere.    I have had negative feedback from engineers outside our team on this exact point, and I agree with them 100%.    So does Greg.    Having this as a GOLDEN RULE for our coding standard casts a very poor light on the standard.    I think we should change it.    We don't need to "fix" all the code immediately.

2. Explicitly listing authors and maintainers PER FILE is unnecessary and egotistical.    It is a maintenance issue as well.    We should stop doing this, except where maintaining existing author information, or author information from files copied from other projects.

3. Leading underscores are, I believe, reserved for compiler and library authors.   Also, there should be no need for any prefixing if we are "m_" prefixing member variables - the most common source of naming clash.

gcolvin commented:
The C++11 Standard says
- Each identifier that contains a double underscore __ or begins with an underscore followed by an uppercase letter is reserved to the implementation for any use.
- Each identifier that begins with an underscore is reserved to the implementation for use as a name in the global namespace

I don't like to do it, but since method parameters are not in the global namespace it's technically OK to use underscores.

gcolvin commented:
Insisting on leading tabs I can live with.  It means you never know how far your code is going to be indented outside of your own environment, and that the column numbers in compiler error messages are wrong, but all our code is already that way, and some people love the tabs for their own reasons.  But insisting on no use of space at all for formatting should be dropped, as vertical alignment done with spaces still looks fine when the tab stops change, but done with tabs it falls apart.

gcolvin commented:
And since these discussions can go on forever to no avail, I'm happy to let Christian decide, embody the decisions in an astylerc, integrate astyle with our workflow, and get on with it.

bobsummerwill commented:
I'm looking at **astyle** now.

Have we ever had any automation of **astyle**, @chriseth?    And/or have developers had integrations in their IDEs in some automated flow.     Have you?

Looks like it does in-place reformatting (unless there is some "report only" mode which would just report IF it would do anything, but not actually do it, which could be automated and reject PRs if running **astyle** would result in any changes).

Not sure I would be overly happy with a "fix it for you automatically" mode, or at least not until we have a canonical POST commit automated build, to ensure that everything still works and that we didn't get any issues introduced by the automated step running amuck.

chriseth commented:
Coding style is not too much about being logical but much more about being consistent across the codebase. So changing some point in some parts is much worse than not changing it, I think.

Furthermore, the variables prefix have clear advantages, especially if you don't have a fully-fledged IDE at hand, e.g. during reviews in github.

Tabs/spaces: It is like it is, while changing tabs in the whole codebase is doable in an automated fashion, I think we should not waste our time on that.

Authors: We did not follow that advice anyway, so I'm fine with removing that.

Automated astyle never worked for me. In general, I think you cannot automate things like coding style, it is just like a natural language dialect and I think AI is not that far yet.

chfast commented:
In general I'm for the changes.

Our coding style really sucks because it cannot be verified by any existing tool designed for that. And we usually waste a lot of time discussing whenever the PR code obeys coding style or not. 

Let's configure `astyle`, `clang-format` or whatever, submit the config to repo and use that tool for new code.

bobsummerwill committed:
Added  "C++ Core Guidelines" to reading list, as per Pawel.
Softened off on the spaces changes back to almost nothing.   Re-added the underscore prefix items.  I don't think it's a good recommendation, but don't care enough to argue more about that.

bobsummerwill commented:
OK - I've revised the language on spaces and tabs.   Please review!

@chriseth - "Furthermore, the variables prefix have clear advantages, especially if you don't have a fully-fledged IDE at hand, e.g. during reviews in github."

Please could you elaborate on what "clear advantages" these prefixes have?
- If parameters were used incorrectly (ie. trying to write to an input-only variable) then you would get build errors assuming the types were set up correctly (ie. const refs, etc).
- In general, parameters are **all** inputs.
- I really hope that we don't have many input/output parameters, because those are bug factories and not a good pattern to be encouraging.
- And our own coding standard says that multiple outputs should be returned using tuples, and to prefer exceptions to error codes.

Also, what advantage does this have:
- Leading underscore "_" to parameter names (both normal and template).

chfast commented:
For more reading I would add also https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md.

bobsummerwill commented:
Done, @chfast!

gcolvin commented:
Leading tabs aren't worth changing.  Some teams (and even languages) insist on them.  Other teams (and even Fortune 100 companies) abhor them.  But the "no spaces for alignment" rule is truly counterproductive -- any attempt at vertical alignment looks horrible in Github or anywhere without 4-space tabs.  So that's the only change I'd insist on.
   I was surprised to find the loophole (only one underscore, only before lowercase, and not in global namespace) in the standard that allows for our leading underscores.  I still think it's asking for trouble, but don't care enough to argue.

gcolvin commented:
To the extent that our style rules can be automated (which is a lot) I think they should be.  I'd much sooner have astyle fix things than waste time training my fingers not to type K&R like they have for 30-some years, and poring over my code looking for misplaced braces and spaces.   It's the mechanical aspects that give the codebase a uniform feel, and make it easier for outside contributors to get oriented.  And automating the mechanical aspects lets code review be about architecture and performance instead.  Bob and I have been discussing the mechanical details, but think we need a way to find out whether code breaks the rules, to fix code that doesn't, and to force code to do so at some point in the PR process.

bobsummerwill commented:
So @gcolvin.

Looking at the diffs as they currently stand ... are you happy with the new language as written?    Or would you strike the "If you have run-on lines, indent as you would for a block." line entirely?

I've re-added the main part of underscores for parameters.   I also agree that it is unnecessary and asking for trouble, but if it is NOT in breach of the C++ standard then I don't care to argue for that any further.     I've left the o_ and io_ exceptions as being removed, because I think those are not at all useful.

chriseth commented on CodingStandards.txt:
```diff
@@ -91,8 +88,6 @@ All other entities' first alpha is lower case.
 4. Variable prefixes:
 
 a. Leading underscore "_" to parameter names (both normal and template).
-- Exception: "o_parameterName" when it is used exclusively for output. See 6(f).
```
I would like to keep that. It is extremely rare and should be flagged as such.

bobsummerwill replied:
Why not flag this extremely rare pattern in the parameter name, like `outputParameterName` rather than having an invented special naming convention?

Or indeed, why bother having a standard at all for something which is extremely rare?  Such things should just have extensive comments, so that they are really noticeable.   Not just noticeable because they use a weird variable prefix.

bobsummerwill replied:
Looks like there are quite a lot of `io_` instances, actually.   Some might be I/O classes, but seeing nearly 200 instances of that string within webthree-umbrella.

gcolvin replied:
Or, being so rare, not worry about it?  My style is that input parameters should be values or const references, and output parameters should be pointers, and pointers should be prefixed with p (my only prefix rule) so it's obvious when I'm dealing with an output parameter.  And since all but the lowest-level code should avoid raw pointers, output parameters indeed should be rare.

chriseth replied:
Prefix is better because you will notice at every point where they are used instead of just their definition.

bobsummerwill commented:
RE:  Automation of style check/fix:

NEW ISSUE - https://github.com/ethereum/webthree-umbrella/issues/486.
